ts-client/src/amm/index.ts

```diff
@@ -0,0 +1,935 @@
+import { AnchorProvider, Program, BN, BorshCoder, Idl } from '@project-serum/anchor';
+import {
+  PublicKey,
+  Connection,
+  Cluster,
+  Transaction,
+  TransactionInstruction,
+  AccountInfo,
+  ParsedAccountData,
+  SYSVAR_CLOCK_PUBKEY,
+} from '@solana/web3.js';
+import { TokenInfo, TokenListProvider } from '@solana/spl-token-registry';
+import { AccountLayout, MintLayout, Token, TOKEN_PROGRAM_ID, u64 } from '@solana/spl-token';
+import { ASSOCIATED_PROGRAM_ID } from '@project-serum/anchor/dist/cjs/utils/token';
+import { calculateWithdrawableAmount, VaultState } from '@mercurial-finance/vault-sdk';
+import invariant from 'invariant';
+import { AccountsInfo, AmmImplementation, DepositQuote, PoolInformation, PoolState, WithdrawQuote } from './types';
+import { Amm, IDL as AmmIdl } from './idl';
+import { Vault, IDL as VaultIdl } from './vault-idl';
+import {
+  DEFAULT_SLIPPAGE,
+  DEVNET_COIN,
+  ERROR,
+  CURVE_TYPE_ACCOUNTS,
+  PROGRAM_ID,
+  SEEDS,
+  VAULT_PROGRAM_ID,
+  WRAPPED_SOL_MINT,
+} from './constants';
+import { StableSwap, SwapCurve, TradeDirection } from './curve';
+import { ConstantProductSwap } from './curve/constant-product';
+import {
+  calculateMaxSwapOutAmount,
+  calculateSwapQuote,
+  computeActualDepositAmount,
+  calculatePoolInfo,
+  getMaxAmountWithSlippage,
+  getMinAmountWithSlippage,
+  getOrCreateATAInstruction,
+  unwrapSOLInstruction,
+  wrapSOLInstruction,
+  getDepegAccounts,
+} from './utils';
+
+type AmmProgram = Program<Amm>;
+type VaultProgram = Program<Vault>;
+
+type Opt = {
+  cluster: Cluster;
+};
+
+const getPoolState = async (poolMint: PublicKey, program: AmmProgram) => {
+  const poolState = (await program.account.pool.fetchNullable(poolMint)) as PoolState;
+  invariant(poolState, `Pool ${poolMint.toBase58()} not found`);
+
+  const account = await program.provider.connection.getTokenSupply(poolState.lpMint);
+  invariant(account.value.amount, ERROR.INVALID_ACCOUNT);
+
+  return { ...poolState, lpSupply: new BN(+account.value.amount) };
```

Comment:
i think BN accept string type number, no need to `+` cast it to number and having the risk of precision loss